### PR TITLE
Implementation of range scans from RFC-0005

### PR DESCRIPTION
**Pull request by hachikuji**
This patch contains the initial implementation of range scans as defined by https://github.com/slatedb/slatedb/blob/main/rfcs/0005-range-queries.md.

### CODE REVIEW

**Review by rodesai — COMMENTED** [review]
Still reviewing the tests, but here's some thoughts after my first pass over the code. Overall this is looking great!

File: src/range_util.rs
Comment (rodesai) [inline src/range_util.rs:470]: How do you think we should handle failures from these, because as I understand it they aren't deterministic. So this means that the test may fail randomly on a pr branch or main. We probably want to update the CI automation to check in the regressions file so that when a failure is found it starts failing consistently.

**Review by criccomini — COMMENTED** [review]
This looks great. Left a few comments here and there. In general, I think there should be some more tests, but I'm guessing you're holding off until things cement a bit more.

File: src/db_iter.rs
Comment (hachikuji) [reply src/db_iter.rs:75]: Initially I had the seek API take a `Bound<Bytes>` as the argument. But it felt weird to allow seeking to `Unbounded` and I thought 99% of use cases probably just wanted `Included(Bytes)`, so I simplified the API and forgot to update the docstring.

File: src/config.rs
Comment (hachikuji) [reply src/config.rs:215]: Makes sense. How about `read_ahead_bytes` to be a little more concise?

File: src/config.rs
Comment (hachikuji) [reply src/config.rs:225]: Sounds good. I'll file an issue to revisit.

File: src/merge_iterator.rs
Comment (hachikuji) [reply src/merge_iterator.rs:52]: I guess it should probably be an error. It might require some additional bookkeeping on the iterator implementations though to keep track of the last returned key. Let me see what this looks like..

File: src/merge_iterator.rs
Comment (hachikuji) [reply src/merge_iterator.rs:200]: Makes sense. I will fix this and add some test cases.

File: src/db.rs
Comment (hachikuji) [reply src/db.rs:1495]: Yeah, all of that makes sense. Getting proptest to play nicely with async is the catch. Let me see if I can find a nicer pattern which also give us access to the seed.

Comment (hachikuji) [reply src/db.rs:1479]: I think I found a reasonable pattern here. All of the inputs are driven by a random seed. In the case of a failure, `proptest` will write the seed to proptest-regressions. For example, if one of the `Db` tests fails, then we might see the following under `proptest-regressions/db.txt`:
```
cc 144db4eba8acd230ccf575391873083d60889043365308de876702bd6ac0236f # shrinks to BytesRange { start_bound: Included(b""), end_bound: Included(b"") 
```
Then we can take that seed and use it directly in the failing test to reproduce the failure.

File: src/block_iterator.rs
Comment (hachikuji) [reply src/block_iterator.rs:67]: Yeah, good call. This is a nice simplification.

File: src/range_util.rs
Comment (hachikuji) [reply src/range_util.rs]: I think the idea is that the regressions file should get checked in to the source repository. Then proptest automatically runs the tests against it.

File: src/db_state.rs
Comment (hachikuji) [reply src/db_state.rs:167]: Yeah, this check was wrong. I think what I intended was `idx + 1 < ssts.len`. Let me write some tests for this.

File: src/mem_table.rs
Comment (hachikuji) [reply src/mem_table.rs]: Basically I couldn't come up with a nice lazy approach which didn't rely on self-referencing data structures (such as through ouroborus which is used by minilsm). I discussed with @rodesai and it seemed simpler for the initial work to filter memory tables up-front. 

I'm definitely open to suggestions here. It's probably possible to refactor the API so that `scan()` returns some kind of result struct which can then be used to build the iterator. For example:
```rust
let scan_result = db.scan(..).unwrap();
let scan_iter = scan_result.iter();
while Some(record) = iter.next().await? {
...
}
```
I think this would work because the iterator can refer to the data structures in the scan result, but it looks a little clumsier.

File: src/db.rs
Comment (hachikuji) [reply src/db.rs:129]: We had that discussion about whether the range to `scan` should be by reference (`scan(&..)` vs `scan(..)`). Since we decided to pass by value, we need `DbIterator` to own the range value (through `SstIterator`) so that it doesn't get dropped when `scan` returns. We might be able to do the same `AsRef` trick that is used for `SsTableHandle`. I will give it a shot. If there is a more straightforward approach, let me know. I'm still pretty new to rust.

File: src/mem_table.rs
Comment (criccomini) [reply src/mem_table.rs]: Yea, let's keep it simple. If it gets in the way we can revisit.

File: src/db.rs
Comment (hachikuji) [reply src/db.rs:129]: @criccomini I think a better approach might be to pull the range/key out of `SstIterator`. I have a draft here which gives the idea: https://github.com/hachikuji/slatedb/blob/range-scans-externalize-sst-iter-range/src/sst_iter.rs#L289. If this looks reasonable, we might want to consider moving it to a follow-up since this patch is already quite large.

**Comment by hachikuji** [comment]
@rodesai @criccomini This is probably read for another look. I think there are two unresolved comments:
- Enforcing range bounds for internal implementations of `SeekToKey`: https://github.com/slatedb/slatedb/pull/342#discussion_r1852919836
- Avoid key copy in `Db::get`: https://github.com/slatedb/slatedb/pull/342#discussion_r1854777437

If it seems reasonable, I'd lke to address these in a follow-up. Happy to do it here as well, but the patch is getting cumbersome to work with.

File: src/db.rs
Comment (criccomini) [reply src/db.rs:129]: Ah, nice. Yea, this looks good. I agree that this should be implemented as a follow-on. As you said, the current PR is pretty lengthy.

**Comment by criccomini** [comment]
LGTM! I think we can commit after rebase.

File: src/bytes_range.rs
Comment (FiV0) [inline src/bytes_range.rs:8]: Do you want to make this public?

Comment (FiV0) [reply src/bytes_range.rs:8]: Nevermind, was playing around with this.

File: src/bytes_range.rs
Comment (rodesai) [inline src/bytes_range.rs:207]: you could avoid this filter by defining some strategy like:
```
pub(crate) fn nonempty_intersecting_ranges(size: usize) -> impl Strategy<Value = (BytesRange, BytesRange)> {
        (arbitrary::rng(), nonempty_range(size))
            .prop_flat_map(|(mut rng, range1)| {
                let start = sample::bytes_in_range(&mut rng, &range1);
                let end_range = BytesRange::new(Included(start.clone()), Unbounded);
                let end = sample::bytes_in_range(&mut rng, &end_range);
                prop_oneof![
                    (Just(range1.clone()), Just(BytesRange::new(Included(start.clone()), Included(end.clone())))),
                    (Just(range1.clone()), Just(BytesRange::new(Unbounded, Included(end.clone())))),
                    (Just(range1.clone()), Just(BytesRange::new(Included(start.clone()), Unbounded))),
                    (Just(range1.clone()), Just(BytesRange::new(Unbounded, Unbounded))),
                ]
            })
    }
```

File: src/bytes_range.rs
Comment (rodesai) [inline src/bytes_range.rs:170]: probably good to unit test the `is_empty` fn

File: src/proptest_util.rs
Comment (rodesai) [inline src/proptest_util.rs:24]: its deterministic in the sense the seed will be recorded by proptest so we can always reproduce the test?

**Review by rodesai — COMMENTED** [review]
This is looking great! Mostly fairly minor comments inline. I'm still working on reviewing all the tests

File: src/db.rs
Comment (rodesai) [inline src/db.rs:1580]: it would probably be good to set a long wal flush interval on this one to ensure we're reading nondurable rows

File: src/sorted_run_iterator.rs
Comment (rodesai) [inline src/sorted_run_iterator.rs:251]: would be good to add some coverage for seek (e.g. seek within the current sst, seek across ssts, seek out of range)

**Review by rodesai — COMMENTED** [review]
Took a pass over the tests. Mostly minor feedback inline

File: src/sst_iter.rs
Comment (rodesai) [reply src/sst_iter.rs:280]: Sure, but this fn is just seeking within an sst. So in the case that the key is out of range it should just exit in the loop below. I assumed this check was just to optimize away the fetches when the key is out of range.

File: src/db.rs
Comment (hachikuji) [reply src/db.rs:1489]: Initially dumping the seed to console would probably be sufficient. Sadly it doesn't look like proptest does this by default. Let me see if I can find a way to do this.

File: src/sst_iter.rs
Comment (hachikuji) [reply src/sst_iter.rs:280]: I think my intention was to skip blocks if we know they are not needed, but the `range_covers_key` is just checking whether `next_key` is greater than the first key, which we know is true. I'll remove the check and we can do a proper optimizaiton in https://github.com/slatedb/slatedb/issues/394.

File: src/db_state.rs
Comment (hachikuji) [reply src/db_state.rs:161]: Not sure why I did that. Probably still getting used to rust borrowing rules...

File: src/proptest_util.rs
Comment (hachikuji) [reply src/proptest_util.rs:24]: Yes, will clarify in the comment

File: src/proptest_util.rs
Comment (hachikuji) [reply src/proptest_util.rs:401]: Good catch.

File: src/sorted_run_iterator.rs
Comment (hachikuji) [reply src/sorted_run_iterator.rs:261]: Yes, good call. I added a test case and fixed a bug in the implementation.

File: src/sorted_run_iterator.rs
Comment (rodesai) [reply src/sorted_run_iterator.rs]: Should we include this in #394 ?

File: src/bytes_range.rs
Comment (rodesai) [inline src/bytes_range.rs:17]: you shouldn't need this check anymore.

**Review by rodesai — APPROVED** [review]
I think we're ready to merge here! Let's :shipit:

**Comment by rodesai** [comment]
Ah I spoke too soon @hachikuji. Let's fix the merge conflicts and then we should be good to go.

**Review by flaneur2020 — APPROVED** [review]
looks great!

**Comment by criccomini** [comment]
🔥

**Pull request merged**
